### PR TITLE
Fancy lists, task list and bunch of pandoc extensions

### PR DESCRIPTION
**Pull request by Omikhleia**
Greetings,
The SILE typesetter has been using (its own local version of) **lunamark** as one of the available options to process Markdown. Besides being based on an old version, it was in a pretty broken shape... As part of an attempt to revive "native Mardown support" in the SILE with decent extensions, I am proposing a few changes to lunamark, which I am therefore relaying upstream here.

In other terms, this PR contains the following things:
- The very first commit is a very small change so that a writer can override the text-only output and process the AST ropes on its own. _Rationale:_ SILE currently uses it to directly hook its custom writer onto its own AST (and thus avoids outputting to text and re-parsing all of it...).
- Subsequent commits consists in a bunch of extensions to the reader:
  - Under the `task_list` option
     -  GitHub-Flavored Markdown's task lists as first-class citizens.
  - Under the `fancy_lists` option
    - Similar to the Pandoc option by the same name, i.e. offering the ability to use a roman number or a letter as list enumerators (in addition to the standard digit), and to use the closing parenthesis as delimiter (in addition to the period), e.g. besides the normal `1.`, one can use `i.`, `a.`, `a)`, etc.
    - Obviously, it plays well too when the `startnum` option is enabled
    - Writers may then honor it (or not), but at least these lists can be processed (even if rendered as usual lists).
  - Under the `pandoc_extensions` option:
    - Strikethrough: `~~deleted~~`
    - Subscripts and superscripts, e.g. respectively `H~2~O`, `2^10^`
    - Inline spans with attributes: `[text]{.class .other key=value key2="value2" ... }`
    - Block divs with attributes, via fenced colon blocks: `::: {.class key=value ... } ...`
    - Attributes also accepts on images (` and on fenced block codes (when the corresponding option for them is enabled)
    - Raw inlines and raw blocks: `{=xxx ...}` after a simple inline code (i.e. backticks) or as "extended" infostring on a fenced code block (idem, when those are enabled) -- content skipped by default, but writers can use them to pass-through direct code to their underlying formatter.
- On the way, I also leveraged the HTML and HTML5 writers to support some of these features. (Something similar could be done possibly for other writers, but I don't know their targets well enough...).

There are other features I'd possibly want to add at a later point, but it would be great and really cool, if this PR is considered and eventually makes it (in that form or another). Moreover, that would also help SILE removing its local copy/fork of the library and use the official package again! :+1: 

Thanks in advance for your time and consideration.

_Reference:_ https://github.com/sile-typesetter/sile/pull/1481
(Just for the record. It is (mostly) the same thing, plus the in-progress changes for SILE itself...)

### CODE REVIEW

**Comment by Omikhleia** [comment]
N.B. Extension-related features currently left out for separate consideration... so you know what could be my wish-list after all the above stuff... In order of importance (from my viewpoint):
- Attributes on headers, e.g. 
  ```
  # My chapter {-}
  ## Ma section {.unnumbered custom-style=Fancy lang=fr}
  # Some chapter {#alternate-hash-for-links}
  ```
- Line blocks (= Pandoc's `line_blocks`)
  ```
  | Typically for some verse
  |  With indentation support
  |   As shown here
  ```
- Some tables.. see #31 
- Example lists (Pandoc's `example_lists`), a.k.a. continuous list with `(@)`
- Nesting fenced divs as in Pandoc. The current proposed implementation doesn't support correct nesting as with Pandoc's `fenced_divs` option. Actually I didn't know these could even be nested, I always used the most basic form such as
  ```
  ::: {lang=fr}
  > Some quote in French
  > Lorem
  :::
  ```
  So I went for that low-hanging fruit which already helps covering a lot, but a better implementation would eventually be welcome.

**Comment by Witiko** [comment]
> - Attributes also accepts on images ( and on fenced block codes (when the corresponding option for them is enabled)

@Omikhleia Do I understand it correctly that this PR implements the image part of [the Pandoc's `link_attributes` extension][1] and not the link part?

 [1]: https://pandoc.org/MANUAL.html#extension-link_attributes

**Comment by Omikhleia** [comment]
@Witiko 
> Do I understand it correctly that this PR implements the image part of [the Pandoc's `link_attributes` extension](https://pandoc.org/MANUAL.html#extension-link_attributes) but not the link part?

That's correct, the `#id` linking hash is not here yet -- the parser only supports classes and key-value pairs for now, in that order.

The hash part could easily be added, though. That will also be needed anyway for header attributes (as in my wish list above)... But again, I first went for the easy-to-reach "low-hanging fruits", in that case the ability to pass e.g. a `width` and `height` to the processor, to size the images accordingly. I will want eventually to have a look at linking, but my cross-reference package proposal for SILE is still a pending PR ;-)

(Besides, the library has not seen patches and releases from some time now... I have no idea if @jgm still maintains it?)

**Comment by jgm** [comment]
I'll defer to @Witiko who is the most active contributor.

**Comment by Witiko** [comment]
@jgm Not a problem. I will want to port these changes to witiko/markdown, so a code review is the logical first step.

**Comment by Witiko** [comment]
> That's correct, the `#id` linking hash is not here yet -- the parser only supports classes and key-value pairs for now, in that order.

I see. My question was mainly about attributes on links (i.e. `[a hypertext link](https://hypertext.link){.some-class}`) as opposed to just images, but I will see after the review.

File: bin/lunamark
Comment (Witiko) [inline bin/lunamark:278]: I see several issues with having multiple extensions grouped into a single flag:

- The name `pandoc_extensions` is not very descriptive.
- There is no guarantee that the syntax extensions will not change in the future, which would forward-compatibility.
- It seems preferable to put the user through the hassle of specifying several options over giving him a single option that does not allow the user to opt out of syntax extensions that they don't want.

Please, split `pandoc_extensions` into the `strikeout`, `superscript`, `subscript`, `fenced_divs`, `bracketed_spans`, `raw_attribute`, `link_attributes` and `fenced_code_attributes` syntax extensions, [see also Pandoc's documentation][1]. Then, document the missing bits of the syntax extensions that are yet to be implemented, such as `#id`s in the `*_attributes` extensions as you mention in https://github.com/jgm/lunamark/pull/36#issuecomment-1187722927 and the lack of support for indirect images and direct/indirect links in the `link_attributes` extension, [see Pandoc's documentation][2].

 [1]: https://pandoc.org/MANUAL.html
 [2]: https://pandoc.org/MANUAL.html#extension-link_attributes

File: lunamark/reader/markdown.lua
Comment (Witiko) [inline lunamark/reader/markdown.lua:801]: `^` is listed twice.

P.S.: Food for thought: should even disabled syntax extensions contribute to `larsers.specialchar`? This seems to make lunamark both backward- and forward-incompatible.

Comment (Witiko) [reply lunamark/reader/markdown.lua:801]:
```suggestion
    larsers.specialchar       = S("*_~`&[]<!\\'\"-.@^")
```

File: lunamark/reader/markdown.lua
Comment (Witiko) [inline lunamark/reader/markdown.lua:803]:
```suggestion
    larsers.specialchar       = S("*_~`&[]<!\\-@^")
```

File: lunamark/reader/markdown.lua
Comment (Witiko) [inline lunamark/reader/markdown.lua:1100]: > Support additional attributes

Consider supporting other forms than just direct images, such as indirect images and direct/indirect links, see [Pandoc's documentation][1]. Hm, I wonder if Pandoc's `link_attributes` also supports attributes on autolinks, i.e. `<https://foo.bar/>{.class key=val}`?

 [1]: https://pandoc.org/MANUAL.html#extension-link_attributes

File: lunamark/reader/markdown.lua
Comment (Witiko) [inline lunamark/reader/markdown.lua:1174]: Can we do this without nested parsers?

File: lunamark/util.lua
Comment (Witiko) [inline lunamark/util.lua:228]: I suggest changing `2` to `_to_` for consistency with existing functions, see e.g. `util.rope_to_string()`.

Comment (Witiko) [reply lunamark/util.lua:228]:
```suggestion
function M.roman_to_number(roman)
```

File: lunamark/util.lua
Comment (Witiko) [inline lunamark/util.lua:253]:
```suggestion
function M.order_to_numberstyle(listprefix)
```

File: lunamark/reader/markdown.lua
Comment (Witiko) [inline lunamark/reader/markdown.lua:1586]: Please explain.

**Review by Witiko — CHANGES_REQUESTED** [review]
Overall, looks very good! However, `pandoc_extensions` needs to be split into separate syntax extensions. Furthermore, it seems necessary to have unit tests for the new syntax extensions added in this pull request (current unit tests [are passing][1]).

 [1]: https://github.com/Omikhleia/lunamark/actions/runs/2712402271

File: lunamark/reader/markdown.lua
Comment (Omikhleia) [reply lunamark/reader/markdown.lua:1586]: If this is accepted, maybe the `merge()` name is ill-chosen, though, and not that self-explaining.  Perhaps simply `output`, or even `ropes_to_output` ?

File: lunamark/reader/markdown.lua
Comment (Omikhleia) [reply lunamark/reader/markdown.lua]: Done :)
Documenting that missing bit would have been longer that actually adding it.
Also modified the html/html5 writer to ouput it when present.

File: bin/lunamark
Comment (Omikhleia) [reply bin/lunamark]: Done.
- `#id` added (see below)
- Options split as distinct flags as requested + missing bits documented
- On the way, I noticed Pandoc allowed escaped spaces in super/subscripts (for some good reasons), so I added that rule too.

File: lunamark/reader/markdown.lua
Comment (Omikhleia) [reply lunamark/reader/markdown.lua]: Done

File: lunamark/reader/markdown.lua
Comment (Omikhleia) [reply lunamark/reader/markdown.lua]: Done

File: lunamark/reader/markdown.lua
Comment (Omikhleia) [reply lunamark/reader/markdown.lua:1160]: Not done yet. Maybe split as a new issue for later? For now I just documented it (in the readers comments and the lunamark(1) help)

File: lunamark/reader/markdown.lua
Comment (Omikhleia) [reply lunamark/reader/markdown.lua:1234]: I haven't found a clear way yet (even trickier now with the split options)

File: lunamark/util.lua
Comment (Omikhleia) [reply lunamark/util.lua]: Done (here and when used)

File: lunamark/util.lua
Comment (Omikhleia) [reply lunamark/util.lua]: Done (here and where used)

File: lunamark/reader/markdown.lua
Comment (Omikhleia) [reply lunamark/reader/markdown.lua]: For now, I opted for `rope_to_ouput` (singular), since the default just calls `util.rope_to_string`)

**Comment by Omikhleia** [comment]
I have applied most of the requested changes.
Now looking at unit tests system ;)

**Comment by Omikhleia** [comment]
Some tests added - let me know if it meets your expectations.
(I also rebased my previous commits, fixing an issue I had introduced with the hash-id and an edge case I had overlooked (revealed by the test, actually - I had not checked Divs without attribute^^)

**Review by Witiko — APPROVED** [review]
Looks good to me, thanks again for all your hard work. In 886b14c, I added style checking to the CI and in d8ec95b, I fixed [all warnings emitted by the style check][1]. This should help us with quality control in future pull requests.

@jgm I also added `lunamark-0.6-1.rockspec` and updated `changelog`. After the merge, can you please update `not yet released` in `changelog`, tag the current commit as 0.6.0, and publish version 0.6.0 to Luarocks?


 [1]: https://github.com/Omikhleia/lunamark/runs/7509907915?check_suite_focus=true

File: lunamark/reader/markdown.lua
Comment (Witiko) [inline lunamark/reader/markdown.lua:1125]:
``` suggestion
  larsers.Span   = parsers.between(parsers.Inline, parsers.lbracket, parsers.rbracket)
                 * parsers.attributes
                 / writer.span
```

File: lunamark/reader/markdown.lua
Comment (Witiko) [inline lunamark/reader/markdown.lua:1130]:
``` suggestion
  larsers.Subscript   = parsers.between(larsers.subsuperscripttext, parsers.tilde, parsers.tilde)
                      / writer.subscript
```

File: lunamark/reader/markdown.lua
Comment (Witiko) [inline lunamark/reader/markdown.lua:1133]:
``` suggestion
  larsers.Superscript = parsers.between(larsers.subsuperscripttext, parsers.circumflex, parsers.circumflex)
                      / writer.superscript
```

**Review by Witiko — COMMENTED** [review]
Fixed a number of style issues as I am porting the `bracketed_spans` syntax extension in https://github.com/Witiko/markdown/pull/207. Let's collate these into a separate large pull request sometime later.

**Pull request merged**
